_posts/2020-02-26-demo-fraud-detection-2.md

```diff
@@ -0,0 +1,192 @@
+
+```java
+public abstract class BroadcastProcessFunction<IN1, IN2, OUT> {
+
+    public abstract void processElement(IN1 value, ReadOnlyContext ctx, Collector<OUT> out) throws Exception;
+
+    public abstract void processBroadcastElement(IN2 value, Context ctx, Collector<OUT> out) throws Exception;
+
+}
+```
+
+The difference is the addition of a `processBroadcastElement` method through which the control messages will arrive. The version of `DynamicKeyFunciton` that allows  modifying the list of data distribution keys at runtime by sending rules definitions to `rules` Kafka topic (`RulesSource`) looks as follows:
+
+```java
+public class DynamicKeyFunction
+    extends BroadcastProcessFunction<Transaction, Rule, Keyed<Transaction, String, Integer>> {
+
+
+  @Override
+  public void processBroadcastElement(Rule rule,
+                                     Context ctx,
+                                     Collector<Keyed<Transaction, String, Integer>> out) {
+    BroadcastState<Integer, Rule> broadcastState = ctx.getBroadcastState(RULES_STATE_DESCRIPTOR);
+    broadcastState.put(rule.getRuleId(), rule);
+  }
+
+  @Override
+  public void processElement(Transaction event,
+                           ReadOnlyContext ctx,
+                           Collector<Keyed<Transaction, String, Integer>> out){
+    ReadOnlyBroadcastState<Integer, Rule> rulesState =
+                                  ctx.getBroadcastState(RULES_STATE_DESCRIPTOR);
+    for (Map.Entry<Integer, Rule> entry : rulesState.immutableEntries()) {
+        final Rule rule = entry.getValue();
+        out.collect(
+          new Keyed<>(
+            event, KeysExtractor.getKey(rule.getGroupingKeyNames(), event), rule.getRuleId()));
+    }
+  }
+}
+```
+
+In the above code `processElement()` receives Transactions and `processBroadcastElement()` receives Rules updates. When a new rule is created, it is distributed and saved in all parallel instances of the operator using `processBroadcastState`, as depicted in Figure 6. We use Rule's ID as the key to store and reference individual rules. Instead of iterating over a hardcoded `List<Rules>`, iteration is performed over entries in the dynamically updated broadcast state.
+
+`DynamicAlertFunction` follows the same logic with respect to storing the rules in the broadcast MapState. As described in Part 1  **[!LINK]**, each message in the `processElement` input is intended to be processed by one specific rule and comes "pre-marked" with a corresponding ID by  `DynamicKeyFunction`. All we need to do is retrieve the definition of the corresponding rule from `BroadcastState` by the provided ID and process it according to the logic required by that rule. At this stage, we will also add messages to the internal function state in order to perform calculations on the required time window of data. We will consider how it is done in the final blog of the series about Fraud Detection.
```

Comment:
add Link!
```suggestion
`DynamicAlertFunction` follows the same logic with respect to storing the rules in the broadcast `MapState`. As described in Part 1  **[!LINK]**, each message in the `processElement` input is intended to be processed by one specific rule and comes "pre-marked" with a corresponding ID created by `DynamicKeyFunction`. All we need to do is retrieve the definition of the corresponding rule from `BroadcastState` and process it accordingly. At this stage, we will also add messages to the internal function state in order to perform calculations on the required time window of data. This will be described in detail in the third and final part of this series on our Fraud Detection case study.
```